File: features/document/src/main/java/com/nivasafinance/features/document/repository/DocumentRepositoryWrapper.java
package com.nivasafinance.features.document.repository;

import com.nivasafinance.features.document.entity.Document;
import com.nivasafinance.features.document.exception.DocumentExceptionFactory;
import lombok.RequiredArgsConstructor;
import org.springframework.context.MessageSource;
import org.springframework.dao.DataAccessException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.UUID;

@Component
@RequiredArgsConstructor
public class DocumentRepositoryWrapper {
    
    private final DocumentRepository documentRepository;
    private final MessageSource messageSource;
    private final DocumentExceptionFactory documentExceptionFactory;
    
    public DocumentRepositoryWrapper(DocumentRepository documentRepository, MessageSource messageSource) {
        this.documentRepository = documentRepository;
        this.messageSource = messageSource;
        this.documentExceptionFactory = new DocumentExceptionFactory(messageSource);
    }
    
    public Document saveWithException(Document document) {
        try {
            return documentRepository.save(document);
        } catch (DataAccessException e) {
            throw documentExceptionFactory.createOperationException("create", e);
        }
    }
    
    public Document findByIdWithException(Long id) {
        return documentRepository.findById(id)
                .orElseThrow(() -> documentExceptionFactory.createNotFoundException(UUID.randomUUID()));
    }
    
    public Page<Document> findAllWithException(Pageable pageable) {
        try {
            return documentRepository.findAll(pageable);
        } catch (DataAccessException e) {
            throw documentExceptionFactory.createOperationException("retrieve", e);
        }
    }
    
    public List<Document> findAllWithException() {
        try {
            return documentRepository.findAll();
        } catch (DataAccessException e) {
            throw documentExceptionFactory.createOperationException("retrieve", e);
        }
    }
    
    public void deleteByIdWithException(Long id) {
        if (!documentRepository.existsById(id)) {
            throw documentExceptionFactory.createNotFoundException(UUID.randomUUID());
        }
        
        try {
            documentRepository.deleteById(id);
        } catch (DataAccessException e) {
            throw documentExceptionFactory.createOperationException("delete", e);
        }
    }
    
    public Document findByIdentifierWithException(UUID identifier) {
        return documentRepository.findByIdentifier(identifier)
                .orElseThrow(() -> documentExceptionFactory.createNotFoundException(identifier));
    }
}

File: features/document/src/main/java/com/nivasafinance/features/document/repository/DocumentRepositoryWrapper.kt
package com.nivasafinance.features.document.repository

import com.nivasafinance.features.document.entity.Document
import com.nivasafinance.features.document.exception.DocumentExceptionFactory
import org.springframework.context.MessageSource
import org.springframework.dao.DataAccessException
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Component
import java.util.UUID

@Component
class DocumentRepositoryWrapper(
    private val documentRepository: DocumentRepository,
    private val messageSource: MessageSource
) {
    private val documentExceptionFactory = DocumentExceptionFactory(messageSource)

    fun saveWithException(document: Document): Document {
        return try {
            documentRepository.save(document)
        } catch (e: DataAccessException) {
            throw documentExceptionFactory.createOperationException("create", e)
        }
    }

    fun findByIdWithException(id: Long): Document {
        return documentRepository.findById(id)
            .orElseThrow { documentExceptionFactory.createNotFoundException(java.util.UUID.randomUUID()) }
    }

    fun findAllWithException(pageable: Pageable): Page<Document> {
        return try {
            documentRepository.findAll(pageable)
        } catch (e: DataAccessException) {
            throw documentExceptionFactory.createOperationException("retrieve", e)
        }
    }

    fun findAllWithException(): List<Document> {
        return try {
            documentRepository.findAll()
        } catch (e: DataAccessException) {
            throw documentExceptionFactory.createOperationException("retrieve", e)
        }
    }

    fun deleteByIdWithException(id: Long) {
        if (!documentRepository.existsById(id)) {
            throw documentExceptionFactory.createNotFoundException(java.util.UUID.randomUUID())
        }

        try {
            documentRepository.deleteById(id)
        } catch (e: DataAccessException) {
            throw documentExceptionFactory.createOperationException("delete", e)
        }
    }

    fun findByIdentifierWithException(identifier: UUID): Document {
        return documentRepository.findByIdentifier(identifier)
            .orElseThrow { documentExceptionFactory.createNotFoundException(identifier) }
    }
}
